DNN Platform/DotNetNuke.Web/Common/WebFormsServiceProvider.cs

```diff
@@ -0,0 +1,36 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information
+
+namespace DotNetNuke.Web.Common
+{
+    using System;
+    using System.Reflection;
+    using System.Web;
+
+    using DotNetNuke.Common;
+    using DotNetNuke.Common.Extensions;
+    using Microsoft.Extensions.DependencyInjection;
+
+    /// <summary>
+    /// Service provider implementation for <see cref="HttpRuntime.WebObjectActivator"/>.
+    /// </summary>
+    internal class WebFormsServiceProvider : IServiceProvider
+    {
+        private const BindingFlags ActivatorFlags =
+            BindingFlags.Instance |
+            BindingFlags.NonPublic |
+            BindingFlags.Public |
+            BindingFlags.CreateInstance;
+
+        /// <inheritdoc />
+        public object GetService(Type serviceType)
+        {
+            var provider = HttpContextSource.Current?.GetScope()?.ServiceProvider;
+
+            return provider != null && (serviceType.IsInterface || serviceType.GetConstructors().Length > 0)
+                ? ActivatorUtilities.GetServiceOrCreateInstance(provider, serviceType)
+                : Activator.CreateInstance(serviceType, ActivatorFlags, null, null, null);
```

Comment:
This is a great improvement and follows the recommendations I made. In https://github.com/dnnsoftware/Dnn.Platform/pull/4098#issuecomment-695230469 you (@GerardSmit) mentioned that this implementation doesn't work for classes that don't have a default constructor implementation. This demonstrates a major short-coming with this recommendation. Unfortunately, it is not an acceptable change because many classes registered with the container will not have a default constructor.

Let's evaluate a concrete example. Given a class called `MyService`
```c#
public class MyService
{
    // omitted implementation
}
```

If you try injecting this class into a WebForms module it will use the Activator and not the Dependency Injection container. This will become a problem if the `MyService` class was registered as a Singleton as it would generate a new instance every time and not re-use the instance as it is supposed to do if registered as a Singleton.

I think we can make some tweaks here to handle these different scenarios
```suggestion
            if (provider != null)
            {
                var service = provider.GetService(serviceType);
                if (service != null)
                {
                    return service;
                }
            }
            
            return Activator.CreateInstance(serviceType, ActivatorFlags, null, null, null);
```

My code suggestion here removes the `ActivatorUtilities` by design. I was recommending it as a nice to have but since WebForms is already heavily dependent on the use of the `Activator` for instantiating non-public constructors of public classes I don't think it will work for us. Here is the idea of my code suggestion
- First attempt to resolve the service and return it if it is not null
- Else use the activator, at this point the code is assuming that it must be activated by the `Activator`

**Concern:** Right now all of these implementations fallback to the Activator. This means dependencies can be resolved being registered with the Dependency Injection container. This scenario will be extremely buggy, and I don't think it is something we should support. 

@GerardSmit  could you give this change or similar a try and run it through the `DotNetBenchmarking` tool. I think this will be faster than using the checks for `IsInterface` and `GetConstructors` as it removes reflection API invocations from every `GetService` API invocation. This is very important to get right because DNN is all built on WebForms and anything we do here will effect **everything** in DNN except Web API controllers. That is a very significant risk/impact that we need to make sure we get right.

cc: @mitchelsellers @bdukes

Reply:
Forgot to include some additional research I did. 

I took a look at the `Activator.CreateInstance` implementation from `dotnet/runtime` project. I was worried the implementations fork depending and they all go through the same code path. This is useful for us, as it doesn't really matter how we invoke the API.

My first idea was to see if we can check the registered services on the container and then decide which code-path to go down. Once the `ServicePovider` is built there is no way to get back to the `ServiceCollection` and there is no additional extension method. It is impossible to retrieve an `IEnumerable` of all registered services. Instead I recommended attempting to resolve the requested `serviceType` and then performing additional rules if necessary.